qiskit/providers/ibmq/jobmanager.py

```diff
@@ -0,0 +1,332 @@
+from datetime import datetime
+from collections import Counter
+from concurrent import futures
+
+from qiskit.circuit import QuantumCircuit
+from qiskit.pulse import ScheduleComponent, Schedule
+from qiskit.compiler import assemble, transpile
+from qiskit.providers.jobstatus import JobStatus
+from qiskit.providers.exceptions import JobError, JobTimeoutError
+from qiskit.result import Result
+from qiskit.qobj import Qobj
+
+from .ibmqbackend import IBMQBackend
+from .utils.decorators import requires_submit
+from .job import IBMQJob
+
+logger = logging.getLogger(__name__)
+
+
+class JobManager:
+    """Job manager for IBM Q Experience.
+
+    Attributes:
+        _executor: executor to handle asynchronous jobs
+    """
+    _executor = futures.ThreadPoolExecutor()
+
+    def __init__(self) -> None:
+        """Creates a new JobManager instance."""
+        self._jobs = []  # type: List[IBMQJob]
+
+        self._future = None
+        self._future_captured_exception = None
+        self._future_error_msg = None
+
+        # Used for caching
+        self._result = None  # type: Optional[List[Union[Result, None]]]
+        self._error_msg = None  # type: Optional[str]
+
+    def run(
+            self,
+            experiments: Union[List[QuantumCircuit], List[Schedule]],
+            backend: IBMQBackend,
+            name_prefix: Optional[str] = None,
+            shots: int = 1024,
+            skip_transpile: bool = False,
+            max_experiments_per_job: Optional[int] = None
+    ) -> int:
+        """Execute a list of circuits or pulse schedules on a backend.
+
+        The circuits or schedules will be split into multiple jobs. Circuits
+        or schedules in a job will be executed together in each shot.
+
+        The circuits will first be transpiled before submitted, using the
+        default transpiler settings. Specify ``skip_transpiler=True`` if you
+        want to skip this step.
+
+        Args:
+            experiments : Circuit(s) or pulse schedule(s) to execute.
+            backend: Backend to execute the experiments on.
+            name_prefix: Prefix of the job name. Each job name will be the
+                prefix followed by an underscore (_) followed by the job index.
+            shots: Number of repetitions of each experiment, for sampling. Default: 1024.
+            skip_transpile: True if transpilation is to be skipped, otherwise False.
+            max_experiments_per_job: Maximum number of experiments to run in each job.
+                If not specified, the default is to use the maximum allowed by
+                the backend.
+                If the specified value is greater the maximum allowed by the
+                backend, the default is used.
+
+        Returns:
+            Number of jobs to be run.
+
+        Raises:
+            JobError: If a job cannot be submitted.
+        """
+        if (all(isinstance(exp, ScheduleComponent) for exp in experiments) and
```

Comment:
Double-checking - `all` or `any`?

Reply:
`assemble` will fail if they're not all the same type, but `any` is probably more efficient here.